### PR TITLE
Add C++, Python, Go code for chapter_hashing

### PR DESCRIPTION
**Pull request by msk397**

### CODE REVIEW

File: codes/python/chapter_hashing/hash_map.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+"""
+File: hash_map.py
+Created Time: 2022-12-14
+Author: msk397 (machangxinq@gmail.com)
+"""
+
+if __name__ == "__main__":
+    """ 初始化哈希表 """
+    Map = {}
```
Comment (krahets) [inline codes/python/chapter_hashing/hash_map.py:9]: Prefer lowercase letters, you could use `mapp`

File: codes/python/chapter_hashing/hash_map.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    """ 添加操作 """
+    # 在哈希表中添加键值对 (key, value)
+    Map[12836] = "小哈"
+    Map[15937] = "小啰"
+    Map[16750] = "小算"
+    Map[13276] = "小法"
+    Map[10583] = "小鸭"
+    print("\n添加完成后，哈希表为\nKey -> Value")
+    for key, value in Map.items():
```
Comment (krahets) [inline codes/python/chapter_hashing/hash_map.py:19]: Please implement `print_dict()` in `print_util.py`

File: docs/chapter_hashing/hash_map.md [hunk 3]
```diff
@@ -63,19 +63,70 @@ comments: true
...
+    # 在哈希表中删除键值对 (key, value)
+    Map.pop(10583)
     ```
 
 === "Go"
 
-    ```go title="hash_map.go"
-
+    ```go title="hash_map_test.go"
+	// 初始化哈希表
```
Comment (krahets) [inline docs/chapter_hashing/hash_map.md:112]: Incorrect indentation

File: codes/go/chapter_hashing/array_hash_map.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	pair := Entry{key: key, val: val}
+	index := a.hashFunc(key)
+	a.bucket[index] = pair
+}
+
+// 删除操作
+func (a *ArrayHashMap) remove(key int) {
+	index := a.hashFunc(key)
+	// 置为空字符，代表删除
+	a.bucket[index] = Entry{}
```
Comment (krahets) [inline codes/go/chapter_hashing/array_hash_map.go:53]: Is null better than an empty `Entry()` object?

File: codes/go/chapter_hashing/array_hash_map.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		}
+	}
+	return pairs
+}
+
+// 获取所有键
+func (a *ArrayHashMap) keySet() []int {
+	var keys []int
+	for _, pair := range a.bucket {
+		if pair.key != 0 {
```
Comment (krahets) [inline codes/go/chapter_hashing/array_hash_map.go:71]: What will happen if there is a pair whose key = 0 ?

**Review by krahets — COMMENTED** [review]
Thanks for the code! Please address the comment.

File: codes/go/chapter_hashing/array_hash_map.go (hunk 5 above)
Comment (msk397) [reply codes/go/chapter_hashing/array_hash_map.go:71]: I thought int type can't be nil, maybe I'm wrong, I'll try your suggestion

Comment (krahets) [reply codes/go/chapter_hashing/array_hash_map.go]: Could `pair` be nil?

File: codes/go/chapter_hashing/array_hash_map.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+	index := key % 100
+	return index
+}
+
+/* 查询操作 */
+func (a *ArrayHashMap) get(key int) string {
+	index := a.hashFunc(key)
+	pair := a.bucket[index]
+	if pair == nil {
+		return "Not Found"
```
Comment (krahets) [inline codes/go/chapter_hashing/array_hash_map.go:37]: Could we return nil here?

Please carefully check the correspondence to the reference Language.

File: docs/chapter_hashing/hash_map.md [hunk 7]
```diff
@@ -63,19 +63,70 @@ comments: true
...
+    """ 删除操作 """
+    # 在哈希表中删除键值对 (key, value)
+    mapp.pop(10583)
     ```
 
 === "Go"
 
-    ```go title="hash_map.go"
-
+    ```go title="hash_map_test.go"
```
Comment (krahets) [inline docs/chapter_hashing/hash_map.md:111]: Just keep the original `hash_map.go`.

File: docs/chapter_hashing/hash_map.md [hunk 8]
```diff
@@ -242,19 +326,146 @@ $$
...
+        void put(int key, string val) {
+            Entry* pair = new Entry(key, val);
+            int index = hashFunc(key);
+            bucket[index] = pair;
+        }
+
+        /* 删除操作 */
+        void remove(int key) {
+            int index = hashFunc(key);
+            // 置为空字符，代表删除
```
Comment (krahets) [inline docs/chapter_hashing/hash_map.md:373]: `// 置为 null ，代表删除`

Please check the comment in other to languages

File: codes/go/chapter_hashing/hash_map_test.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+	/* 添加操作 */
+	// 在哈希表中添加键值对 (key, value)
+	mapp[12836] = "小哈"
+	mapp[15937] = "小啰"
+	mapp[16750] = "小算"
+	mapp[13276] = "小法"
+	mapp[10583] = "小鸭"
+	fmt.Println("\n添加完成后，哈希表为\nKey -> Value")
+	for key, value := range mapp {
```
Comment (krahets) [inline codes/go/chapter_hashing/hash_map_test.go:24]: Please implement `PrintMap()` in Go

**Review by krahets — COMMENTED** [review]
Hi @msk397 , there are a few comments to be addressed. Thanks!

File: docs/chapter_hashing/hash_map.md (hunk 7 above)
Comment (msk397) [reply docs/chapter_hashing/hash_map.md:111]: But in this part, there is no way to separate different hash map operations from test samples, unless the current function is directly called by the test function

Comment (krahets) [reply docs/chapter_hashing/hash_map.md:111]: I see. My fault.

**Comment by krahets** [comment]
By the way, it is recommended to press `resolve` button after fixing.

**Comment by msk397** [comment]
Sorry, this is my first time participating in an open source project, there are a lot of deficiencies.🥲 Thank you for your code review! I will fix these mistakes.

**Comment by krahets** [comment]
> Sorry, this is my first time participating in an open source project, there are a lot of deficiencies.🥲 Thank you for your code review! I will fix these mistakes.

Not a problem! Unfamiliar at first but you get used to it.

File: codes/go/pkg/print_utils.go [hunk 10]
```diff
@@ -96,3 +96,10 @@ func showTrunk(t *trunk) {
 	showTrunk(t.prev)
 	fmt.Print(t.str)
 }
+
+// PrintHashMap Print a hash map
+func PrintMap(m map[int]string) {
```
Comment (krahets) [inline codes/go/pkg/print_utils.go:101]: Please implement a function with Generics, ie., we should not limit the type of key and value.

Comment (msk397) [reply codes/go/pkg/print_utils.go]: I added PrintMap by reference to PrintSlice function. If Generics are required, does PrintSlice function also need to be supported?

**Review by krahets — APPROVED** [review]
Thanks @msk397 for the great work! Now merge it.

**Comment by msk397** [comment]
> I'm wondering which languages you mainly write in. It is recommended to focus on working on the language you are most familiar with.

I will focus on my main language. Thanks for your advice!

**Comment by krahets** [comment]
> I will focus on my main language. Thanks for your advice!

Yes, or push the code in one language at first. And you can fix other languages following the comments. After merging, push other languages.

File: codes/go/pkg/print_utils.go (hunk 10 above)
Comment (krahets) [reply codes/go/pkg/print_utils.go]: Sorry that I missed this comment. Yes, I think it should also support generics. Could you help update it? Thanks!

Comment (msk397) [reply codes/go/pkg/print_utils.go]: No problem! @krahets

**Pull request merged**